src/v/datalake/translation/deps.cc

```diff
@@ -10,12 +10,545 @@
+        // todo: add logic to wait for enough data to translate.
+        // currently we just break even if a single batch of data is available
+        constexpr auto poll_duration = 2s;
+        constexpr auto poll_limit = (poll_duration * 15) - 1s;
+        const auto deadline = ss::lowres_clock::now() + poll_limit;
+        while (!has_more_data_to_translate(last_translated_offset)) {
+            co_await ss::sleep_abortable(poll_duration, as);
+            if (ss::lowres_clock::now() >= deadline) {
+                co_return std::nullopt;
+            }
+        }
+        update_translation_target();
+        auto read_begin_offset = min_offset_for_translation();
+        if (last_translated_offset) {
+            read_begin_offset = kafka::next_offset(
+              last_translated_offset.value());
+        }
+        co_return read_begin_offset;
+    }
+
+    ss::future<std::optional<model::record_batch_reader>> make_log_reader(
+      kafka::offset begin_offset,
+      ss::io_priority_class io_priority,
+      ss::abort_source& as) final {
+        auto max_translatable_offset = max_offset_for_translation();
+        if (
+          !max_translatable_offset
+          || max_translatable_offset.value() < begin_offset) {
+            co_return std::nullopt;
+        }
+        auto log_reader = co_await _partition_proxy->make_reader(
+          {kafka::offset_cast(begin_offset),
+           kafka::offset_cast(max_translatable_offset.value()),
+           0,
+           std::numeric_limits<size_t>::max(),
+           io_priority,
+           std::nullopt,
+           std::nullopt,
+           as});
+        auto tracker = kafka::aborted_transaction_tracker::create_default(
+          _partition_proxy.get(), std::move(log_reader.ot_state));
+        co_return model::make_record_batch_reader<kafka::read_committed_reader>(
+          std::move(tracker), std::move(log_reader.reader));
+    }
+
+    kafka::offset min_offset_for_translation() const final {
+        if (_partition->is_read_replica_mode_enabled()) {
+            return model::offset_cast(_partition_proxy->start_offset());
+        }
+        return model::offset_cast(_partition_proxy->local_start_offset());
+    }
+
+    std::optional<kafka::offset> max_offset_for_translation() const final {
+        // We factor in LSO to ensure only committed transactional batches are
+        // translated.
+        auto lso = _partition_proxy->last_stable_offset();
+        if (lso.has_error()) {
+            return std::nullopt;
+        }
+        return kafka::prev_offset(model::offset_cast(lso.value()));
+    }
+
+    ss::future<std::error_code> update_highest_translated_offset(
+      kafka::offset new_offset,
+      model::term_id term,
+      model::timeout_clock::duration timeout,
+      ss::abort_source& as) final {
+        std::optional<model::timestamp> new_ts{};
+        if (_translation_target && new_offset >= _translation_target->offset) {
```

Comment:
nit: I think this makes sense, but I think it'll be easy to forget the context about why this looks the way it does. I think a comment here would be useful